### PR TITLE
cpp: rewrite?

### PR DESCRIPTION
**Pull request by Spongman**
c++ code that doesn't suck balls, actually implements the lexer functionality and passes the tests (borrowed from the C version).

_zero_ memory allocations btw.

### CODE REVIEW

**Comment by Sirraide** [comment]
Yeah, the current implementation is really just ‘C with classes’ and not C++...

**Comment by arghasen** [comment]
I like this implementation

File: cpp/src/lexer/lexer.cc [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+		case ')': return { TokenType::rparen    };
+		case ',': return { TokenType::comma     };
+		case '+': return { TokenType::plus      };
+		case '{': return { TokenType::lsquirly  };
+		case '}': return { TokenType::rsquirly  };
+	}
+
+	// parse integer
+	if (isdigit(ch)) {
+		while (position_ < length_ && isdigit(input_[position_]))
```
Comment (vgasparyan1995) [inline cpp/src/lexer/lexer.cc:44]: This (and the similar lines below) looks like a good place to use string_view::first_not_of.

File: cpp/src/lexer/lexer.hh [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	lsquirly,
+	rsquirly,
+	function,
+	let,
+};
+
+
+struct Token final
+{
+	const TokenType type;
```
Comment (vgasparyan1995) [inline cpp/src/lexer/lexer.hh:26]: This const seems too harsh. It effectively strips away copy-assignability. As a user, I'd expect the Token returned by nextToken() to be a trivially copyable simple struct. Let me take a const Token if I mean it.

File: cpp/src/lexer/lexer.cc (hunk 1 above)
Comment (Spongman) [reply cpp/src/lexer/lexer.cc:44]: i very much doubt that's more efficient that `is*()`

File: cpp/src/lexer/lexer.hh [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+class Lexer final
+{
+	public:
+		Lexer(std::string_view);
+		Token nextToken() noexcept;
+
+	private:
+		const std::string_view input_;
+		const size_t length_;
+		size_t position_ = 0;
```
[the remark this answers is not in the document]
Comment (Spongman) [reply cpp/src/lexer/lexer.hh:40]: why update two values (start & length) instead of just one (position) ?

File: cpp/src/lexer/lexer.cc (hunk 1 above)
Comment (Spongman) [reply cpp/src/lexer/lexer.cc]: new version uses a `string_view` and  a `string_view::iterator` only.

**Comment by vhladko** [comment]
@nicdgonzalez can you take a look to this one ? :)

**Review by nicdgonzalez — APPROVED** [review]
Looks good to me 👍🏽

**Comment by vhladko** [comment]
Okay, then lets go :)

**Pull request merged**